### PR TITLE
syscalls: print syscall numbers in hex instead of dec to accomodate macos x86 64bit

### PR DESCRIPTION
**Pull request by jacobrosenthal**
You guys jumped the gun on merging my WIP PR #10956

One necessary fix is that mach syscalls were arbitrarily entered from @stek29's script in that PR with swi 0x81. Turns out that they get looked up in get_swi with -1 which turns int 0x80, so all calls have to use 0x80.

This finds yet a few more missing syscalls. (though not all)

### CODE REVIEW

File: libr/core/cmd_anal.c [hunk 1]
```diff
@@ -5073,8 +5073,13 @@ static void cmd_anal_syscall(RCore *core, const char *input) {
 		} else {
 			list = r_syscall_list (core->anal->syscall);
 			r_list_foreach (list, iter, si) {
-				r_cons_printf ("%s = 0x%02x.%u\n",
+				// hack for macos syscalls which are hex 0x1000000 and above and dont care about swi
+				if (si->num < 0x1000000) {
```
Comment (Maijin) [inline libr/core/cmd_anal.c]: Couldn't you test against the value of e asm.os instead?

Comment (jacobrosenthal) [reply libr/core/cmd_anal.c]: I could. How can i get from core->??->asm.os

Comment (Maijin) [reply libr/core/cmd_anal.c]: see `r_config_get (core->config, "asm.os");`

File: libr/core/cmd_anal.c [hunk 2]
```diff
@@ -1220,7 +1222,12 @@ R_API char *cmd_syscall_dostr(RCore *core, int n) {
 		res = r_str_appendf (res, "%d = unknown ()", n);
 		return res;
 	}
-	res = r_str_appendf (res, "%d = %s (", item->num, item->name);
+
+	if (!strcmp (os, "macos")) {
```
Comment (ret2libc) [inline libr/core/cmd_anal.c]: I'm not familiar with this part of the code, but could you tell us why it's necessary to special handling `macos` in all these parts of the code?

Comment (jacobrosenthal) [reply libr/core/cmd_anal.c]: mac uses large number hex syscalls, take a look at the sdb 
https://github.com/radare/radare2/blob/master/libr/syscall/d/darwin-x86-64.sdb.txt#L20

its not intuitive to talk about 536870939 or 536870934 syscall, but rather 0x2000001B or 0x20000016

Comment (ret2libc) [reply libr/core/cmd_anal.c]: can't we just use always hex then?

Comment (jacobrosenthal) [reply libr/core/cmd_anal.c]: Thats up to you all

Comment (jacobrosenthal) [reply libr/core/cmd_anal.c]: Though note the hardcoding of linux's int 80 would still need to be switched on os for other platforms.

File: libr/core/cmd_anal.c [hunk 3]
```diff
@@ -5065,16 +5073,24 @@ static void cmd_anal_syscall(RCore *core, const char *input) {
...
+					}
 				} else {
 					eprintf ("Unknown syscall name\n");
 				}
 			}
 		} else {
 			list = r_syscall_list (core->anal->syscall);
 			r_list_foreach (list, iter, si) {
-				r_cons_printf ("%s = 0x%02x.%u\n",
+				if (!strcmp (os, "macos")) {
```
Comment (ret2libc) [inline libr/core/cmd_anal.c]: why don't you print `si->swi` for macos as well? And why can't we just switch from `%u` to `0x%x` for all architectures?

Comment (radare) [reply libr/core/cmd_anal.c]: strcmps inside a loop is going to slowdown a lot. add an bool isMacOS = !strcmp (os, "macos"); before the rListForeach() and just check for this variable in there

**Comment by jacobrosenthal** [comment]
@ret2libc  Technically should maintain int 80 for 32bit I guess, but 64 bit, which has been modern since 2011 doesnt utilize I guess https://filippo.io/making-system-calls-from-assembly-in-mac-os-x/

**Comment by ret2libc** [comment]
That's a good point. The true though is true for linux 64bit AFAIK.

@sivaramaaa thoughts? what are we doing for linux 64 bits with `syscall` instruction?

**Comment by ret2libc** [comment]
BTW, I think that if some special handling needs to be done for `syscall`-like system calls, it shouldn't be based on the OS as it is now, but on the architecture/type of syscal. AFAIK macos on x86-64 has nothing special in this regard.

**Comment by jacobrosenthal** [comment]
If everyone else agrees to switch to hex printing all syscall numbers then Ill just rebase and remove all ifelses leaving the printing int 80 on 64 bit systems. Seems like a fine compromise

**Comment by jacobrosenthal** [comment]
Formatting goes from
`0x80.10`
to (please vote)
`0x80.0A` or `0x80 0x0A`

**Comment by jacobrosenthal** [comment]
Something else Im trying to solve. Some of these take 8 arguments. We have

https://www.unix.com/man-page/osx/2/getdirentriesattr/In 32fc31e I added more registers to 64 bit, as some in sdb have up to 8 args but I might need to change something somewhere else. Any ideas?
```
[0x01057929]> as 0x1000064
No debug register profile defined for 'A6'.
No debug register profile defined for 'A7'.
0x1000064 = iokit_user_client_trap (0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0xffffffffffffffff, 0xffffffffffffffff)
[0x01057929]> 
```

**Comment by radare** [comment]
Svc 0x80 also applies to ios/arm64

**Comment by radare** [comment]
can you rebase? i think we can merge this now and continue working on another PR

**Comment by jacobrosenthal** [comment]
re int80, Basically Im currently leaving the int (80 or otherwise) untouched in this pr.

This would be ready when I can solve https://github.com/radare/radare2/pull/10985#issuecomment-412222757

**Comment by XVilka** [comment]
Doesn't even build on AppVeyor and 5 [XX] on Travis.

File: libr/syscall/syscall.c [hunk 4]
```diff
@@ -83,9 +83,16 @@ R_API bool r_syscall_setup(RSyscall *s, const char *arch, int bits, const char *
 		s->regs = fastcall_mips;
 	} else if (!strcmp (arch, "avr")) {
 		s->sysport = sysport_avr;
-	} else if (!strcmp (os, "darwin") || !strcmp (os, "osx") || !strcmp (os, "macos")) {
+	} else if (!strcmp (os, "macos")) {
```
Comment (ret2libc) [inline libr/syscall/syscall.c]: Why did you remove darwin and the others?

Comment (jacobrosenthal) [reply libr/syscall/syscall.c]: I added the darwin while goofing around and it got merged. The osx I cant find anywhere else in the codebase and was added as part of the first syscall pull request, I believe incorrectly, then later patched to add macos but osx wasnt removed.

**Comment by jacobrosenthal** [comment]
@XVilka fails mention fastcall.h so maybe related to increasing registers  https://github.com/radare/radare2/pull/10985#issuecomment-412222757

Ill pull that out here to merge for now since I cant get that any further atm...

File: libr/syscall/syscall.c (hunk 4 above)
Comment (jacobrosenthal) [reply libr/syscall/syscall.c]: Looks like we use macos and darwin both internally (but again, no osx) https://github.com/radare/radare2/blob/master/libr/bin/format/mach0/mach0.c#L2120-L2125

I found some random older mac bins and some older stuff (though still 64 bit) did come in as darwin so thats still an exercised code path so Ill actually put my darwin check back in.

The only other use of darwin is in r2 egg? whats thats? Perhaps that file needs an ```|| !strcmp (os, "macos"``` as well??

Comment (jacobrosenthal) [reply libr/syscall/syscall.c]: PRed that here https://github.com/radare/radare2/pull/11090

**Comment by jacobrosenthal** [comment]
Lastest commit gets asc and anything else so I think all interrupt and syscall numbers are printed in hex at least from cmd_anal.c

Id say this is only waiting on understanding more about darwin/macos etc selection in https://github.com/radare/radare2/pull/11090

**Comment by ret2libc** [comment]
I think you should remove from this pr the part related to the if macos/OS X/Darwin and change that in #11090, so that this can be merged easily

**Comment by jacobrosenthal** [comment]
Done. lets merge.

**Comment by ret2libc** [comment]
@radare the reason was that some os have very big numbers for syscalls numbers and it’s easier to understand them in hex ( eg 0x2000001 vs the decimal representation)

**Comment by jacobrosenthal** [comment]
uhhhmm Read above? macos syscalls are offset by 0x2000000.

unreadable
```
exit = 0x80.33554433
fork = 0x80.33554434
read = 0x80.33554435
```

readable
```
exit = 0x2000001
fork = 0x2000002
read = 0x2000003
```


I had if else statements to only do this for macos arch, but people were against those so we compromised on printing swi on mac even though it doesnt care about it at all, but changing all syscalls to hex so we dont have if else everwhere.

**Comment by radare** [comment]
We can format big numbers as hex and small ones as dec. the testsuite would be green and will work for the case. What do you think?

**Comment by jacobrosenthal** [comment]
There ya go

**Comment by jacobrosenthal** [comment]
If I catch you correctly, mallocing a larger string and copying into it and replacing?
That seems.. much more annoying for a little more readability

**Comment by ret2libc** [comment]
What @radare suggested is very different from this. You don’t need to if/else every call but you could just create a helper function which calls r-cons-printf or whatever with the right format string based on the number (and not on the fact you are on OS X)

Moreover you don’t need to print hex when you are printing json, because json is for machines and they can easily understand big decimal numbers.

**Comment by jacobrosenthal** [comment]
Someones going to have to at least psuedocode for me

**Comment by radare** [comment]
```
if (cmd_anal_isMacos64(core)) {
 		res = r_str_appendf (res, "0x%02x = %s (", item->num, item->name);
```

->

```
const char *formatFor(int n, const char *a, const char *b) {
  return n>1000?a: b;
}

res = r_str_appendf (res, formatFor(item->num), item->num, item->name);
```

**Comment by radare** [comment]
for example.

**Comment by jacobrosenthal** [comment]
I like the %s solution better. Implemented.

I kinda dont care. .just posting for posterity. Theres still something else odd in the macosx syscalls from that json list of syscalls we used. it has several arguments with 7 and 8 args in certain syscalls like bsdthread_register

But then radare bitches because there are only 6 int regs for syscalls defined
```
No debug register profile defined for 'A6'.
No debug register profile defined for 'A7'.
```
And I think thats all the regs there are, and indeed the disassembly shows it only using 6 for bsdthread_register 
```
/ (fcn) sym.runtime.bsdthread_register 71
|   sym.runtime.bsdthread_register (int arg_8h);
|           ; arg int arg_8h @ rsp+0x8
|           0x01057e70      488d3db9ffff.  lea rdi, [sym.runtime.bsdthread_start] ; 0x1057e30 ; "L\x89\xc4RQVH\x8d\xb9\x88" ; rdi=0x1057e30 -> 0x52c4894c
|           0x01057e77      48c7c6000000.  mov rsi, 0                  ; rsi=0x0
|           0x01057e7e      48c7c2000000.  mov rdx, 0                  ; rdx=0x0
|           0x01057e85      49c7c2000000.  mov r10, 0                  ; r10=0x0
|           0x01057e8c      49c7c0000000.  mov r8, 0                   ; r8=0x0
|           0x01057e93      49c7c1000000.  mov r9, 0                   ; r9=0x0
|           0x01057e9a      48c7c06e0100.  mov rax, 0x200016e          ; rax=0x200016e
|           0x01057ea1      0f05           syscall                     ; 0x200016e = bsdthread_register (0x01057e30, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000)
```

this list has bsdthread_register with 6 args maybe this is a better list for modern mac. Not a tidy json list to generate from sadly.
https://opensource.apple.com/source/xnu/xnu-1504.3.12/bsd/kern/syscalls.master

For now it may be an imperfect lookup table for a few of the calls.. shrug

Im done here. Merge when ready.

**Comment by jacobrosenthal** [comment]
@wargio  Fixed

File: libr/core/cmd_anal.c [hunk 5]
```diff
@@ -1205,6 +1205,10 @@ static void cmd_anal_trampoline(RCore *core, const char *input) {
 	}
 }
 
+const char *syscallNumber(int n) {
+  return sdb_fmt (n > 1000 ? "0x%x" : "%d", n);
```
Comment (radare) [inline libr/core/cmd_anal.c]: use tabs

File: libr/core/cmd_anal.c [hunk 6]
```diff
@@ -1205,6 +1205,10 @@ static void cmd_anal_trampoline(RCore *core, const char *input) {
 	}
 }
 
+const char *syscallNumber(int n) {
```
Comment (radare) [inline libr/core/cmd_anal.c]: and this must be static

File: libr/core/cmd_anal.c [hunk 7]
```diff
@@ -5086,36 +5091,38 @@ static void cmd_anal_syscall(RCore *core, const char *input) {
 			if (input[2] == ' ') {
 				if (!isalpha (input[3]) && (n = r_num_math (num, input + 3)) >= 0 ) {
 					si = r_syscall_get (core->anal->syscall, n, -1);
-					if (si)
-						r_cons_printf (".equ SYS_%s %d\n", si->name, n);
+					if (si){
```
Comment (radare) [inline libr/core/cmd_anal.c]: missing space before {

File: libr/core/cmd_anal.c [hunk 8]
```diff
@@ -5086,36 +5091,38 @@ static void cmd_anal_syscall(RCore *core, const char *input) {
...
 				}
 				r_list_free (list);
 			}
 		} else {
 			if (input[1] == ' ') {
 				if (!isalpha (input[2]) && (n = r_num_math (num, input + 2)) >= 0 ) {
 					si = r_syscall_get (core->anal->syscall, n, -1);
-					if (si)
-						r_cons_printf ("#define SYS_%s %d\n", si->name, n);
+					if (si){
```
Comment (radare) [inline libr/core/cmd_anal.c]: missing space before {

**Comment by jacobrosenthal** [comment]
done

**Comment by radare** [comment]
merged!
